### ISSUE TITLE
[Feature] Critic Agent: Web Search Support

### ISSUE THREAD

**Issue by xiaofei-zheng** [AMD-AGI/Hyperloom#170]
## Summary

Give Hyperloom's **critic agent** the ability to perform web search, so it can ground its critique in up-to-date external information (latest docs, framework changelogs, known issues, similar bug reports) rather than relying purely on training data.

## Background

The critic agent today reasons only from local context (session, KB, code). For many critique tasks (e.g. "is this kernel optimization aligned with latest sglang/vLLM API?", "is this error a known regression?"), live web information would significantly improve accuracy.

## Scope

- Pluggable web-search tool exposed to critic agent (default: a web-search MCP)
- Query construction from critic context (search prompt template)
- Result ranking + summarization before feeding back to critic reasoning
- Citations: when critic uses search results, include source links in its output
- Cost / rate-limit guardrails (per-session search budget)

## Out of scope

- Building a custom search engine
- Long-term web crawling / indexing (use existing search backend)

## Acceptance

- Critic agent can answer a representative "is X aligned with current sglang API?" question by issuing web search and citing sources, with results visibly improving over the no-search baseline

**Comment by haishuok0525** [comment]
## Implementation status (branch `feature/haiskong/critic-web`)

This comment summarizes the PR #170 implementation, test evidence, and **when web search actually fires** inside the critic agent.

### What landed

- New package `critic-agent/runtime/web_tools/` — pluggable `web_search` / `web_fetch` clients mirroring Primus-Claw Brain env names (`WEB_SEARCH_PROVIDER`, `TAVILY_API_KEY`, `WEB_FETCH_ENABLED`, …).
- `CriticAgentBackend` multi-turn Codex tool loop (`_run_reasoning_loop`) with graceful degradation when tools are disabled or providers are misconfigured.
- `critic.md` — "Web verification" section (when to search, citation rules, error handling).
- Unit + integration tests: **128 passed** locally (`critic-agent/runtime/tests/test_web_tools_*`, `test_critic_agent_tool_loop.py`, existing critic backend tests).

**Default behavior is off** — set `CRITIC_WEB_TOOLS_ENABLED=true` plus a provider key to enable.

---

### When does `web_search` trigger? (timing)

Web search is **not** automatic on every optimizer tick. It only runs inside the critic's **Phase 2 reasoning** step, and only when all gates below pass:

| Gate | Condition |
|------|-----------|
| 1. Coordinator | Orchestration emitted a `propose_action` → proposal is in critic inbox / `pending_proposals` |
| 2. Critic tick | Coordinator `_reactor_pass("critic")` → `CriticAgentBackend.run()` |
| 3. prepare-review | `judge_bundle.proposals` is **non-empty** — if empty, LLM is skipped entirely (`"(skipped — no proposals)"`) |
| 4. Config | `CRITIC_WEB_TOOLS_ENABLED=true` **and** a usable provider (e.g. `WEB_SEARCH_PROVIDER=tavily` + `TAVILY_API_KEY`) |
| 5. Model choice | Codex call uses `tool_choice="auto"` — the model **decides** whether to call `web_search` (prompt guides: uncertain API flags, KB priors mentioning upstream fixes, unverifiable external benchmark claims) |

Flow:

```
propose_action → pending_proposals → critic reactor tick
  → prepare-review → (proposals?) → _reason() → _run_reasoning_loop()
      → [optional] web_search / web_fetch tool rounds (max CRITIC_WEB_MAX_TOOL_TURNS, default 4)
      → review_verdicts JSON → commit-review
```

**Contrast with Primus-Claw Brain:** Brain exposes `web_search` on every user task when `WEB_SEARCH_PROVIDER=tavily`. Critic only searches during **proposal review**, and only if the model chooses to.

---

### Test evidence

#### 1) Local regression (no provider keys)

Smoke marathon on `primus-claw-dev` with `CRITIC_WEB_TOOLS_ENABLED=true`, `WEB_SEARCH_PROVIDER=disabled`, no API keys:

- Optimizer log: `runtime.web_tools [INFO] web tools enabled but neither search nor fetch is usable (provider chain=(), fetch_enabled=False)`
- Critic boot + verdict path unchanged (regression PASS)
- Session: `4527535e-075b-4467-84fb-39f154fca739` (sandbox baseline failed for unrelated sglang/env reasons)

#### 2) Prod Primus-Claw Tavily reference (provider works in platform)

Separate probe on **`primus-claw` prod** namespace (Brain `WebSearchService`, not Hyperloom critic):

- Secret confirms: `WEB_SEARCH_PROVIDER=tavily`, `TAVILY_API_KEY` set, `WEB_FETCH_ENABLED=true`
- Session `b377e9fb-c703-4c0e-8cc4-1846f231cc60`: `web_search` returned real URLs (github.com/sgl-project/sglang/releases, etc.) — Tavily path verified on platform

To run critic-side Tavily in a marathon sandbox, export the same env vars into the optimizer process:

```bash
export CRITIC_WEB_TOOLS_ENABLED=true
export WEB_SEARCH_PROVIDER=tavily
export TAVILY_API_KEY=<from primus-claw-secrets or Tavily dashboard>
# optional: export WEB_FETCH_ENABLED=true
```

---

### Known gaps / follow-ups

- [ ] End-to-end critic `web_search` call with a live Tavily key inside a Hyperloom marathon (blocked on dev sandbox default image pull; prod image override works for Claw Brain probes)
- [ ] Optional: `anthropic_native` provider for critic when using Claude + SaFE gateway (no third-party key) — same pattern as Primus-Claw `executeAnthropicSearch`
- [ ] Acceptance scenario from issue: representative "is X aligned with current sglang API?" with visible citation in `review_verdict.notes[]` — ready once Tavily key is injected into optimizer env

---

### Enable checklist for reviewers

```bash
export CRITIC_WEB_TOOLS_ENABLED=true
export WEB_SEARCH_PROVIDER=tavily   # or serper / brave
export TAVILY_API_KEY=...
export WEB_SEARCH_FALLBACK=          # optional CSV
export WEB_FETCH_ENABLED=false       # off by default
```

Run: `inference_optimizer optimize ... --critic-agent` — watch `<session_dir>/agents/critic/` and optimizer logs for `critic_agent_backend tool-call turn=... tools=['web_search']`.

**Comment by haishuok0525** [comment]
Implementation PR: https://github.com/AMD-AGI/Hyperloom/pull/309

Branch `feature/haiskong/critic-web` — see earlier comment for test results and trigger-timing details.
